### ISSUE TITLE
Unity app is causing too many requests on crafatar.com

### ISSUE THREAD

**Issue by jomo** [crafatar/crafatar#315]
Some application that only identifies itself as `UnityPlayer/2021.3.22f1 (UnityWebRequest/1.0, libcurl/7.84.0-DEV)` is currently causing *a lot* of requests against crafatar.com.

The request logs show that sometimes several thousand requests are caused by this application each second, many of which request the same URL for the same source IP address within the same second.

This is too much for the small VM hosting crafatar.com. If you know what this app / game is, please let me know.
Ideally the developers of this application would chime in here.

You'll have to optimize your code to cause less requests, or host your own instance of crafatar (see README for a docker setup).

**Comment by jomo** [comment]
This seems to be caused by @QuestCraftPlusPlus, most likely their closed-source launcher.

Hopefully the developers can chime in here and fix this soon.

//cc @thejudge156, @CADIndie, @MrNavaStar

**Comment by MrNavaStar** [comment]
I no longer am a part of this project, but the source code that actually implements the api call is here: https://github.com/QuestCraftPlusPlus/Pojlib/tree/QuestCraft

I didn't realize they made the unity part of the project closed...

**Comment by polyzium** [comment]
Yeah, the folks on Discord said that they are selling the launcher source code for $10k to companies. You can make your own, since the bits and pieces are open source.

**Comment by MrNavaStar** [comment]
Well sorry about this! I unfortunately only worked on the back end side of the app (never anything in unity) and It seems that unity is calling the function, wish I could do more to help! In the meantime if Judge or cad don't respond I would just block the user agent, I understand that the extra must be frustrating.

**Comment by jomo** [comment]
Thanks for chiming in @MrNavaStar! The User Agent is currently being redirected to this issue :)
Unfortunately this also affects any other (unrelated) Unity apps, but it's the best I can currently do.

**Comment by CADIndie** [comment]
Hello, Lead Project Manager here 👋

We are really really sorry for this, we tossed our craftar implementation in fairly quickly and I for some reason never got an email for this thread. We will work to fix this issue ASAP and we will make sure to make it identifiable who is using QuestCraft.

Please reach out to us inside our Discord at discord.gg/QuestCraft or DM at cadindie again in Discord.

Again, we are really sorry for the inconvenience.

**Comment by polyzium** [comment]
Until further response I am blacklisting crafatar.com on my side in order to avoid being blamed. I am getting tired of this.

**Comment by jomo** [comment]
Thanks for taking care of this @CADIndie!

@polyzium there is no need to block crafatar on your side. Requests are currently redirected, so they don't cause any load and crafatar.com is running stable. Nobody is blaming users and I have no reason to believe this was anything but an honest mistake.

**Comment by CADIndie** [comment]
@jomo Hello, could you please confirm whether the majority of the requests are coming from one IP or if it's multiple? We are trying to understand what went wrong since we have made no changes whatsoever between the dates you show peak usage and therefore this shouldn't be an error on our side. Furthermore we have utilized this same system for millions of downloads since at least March of this year once again making us believe this can't be an issue on our end.

**Comment by jomo** [comment]
It's multiple IPs (v4 and v6 from different parts of the world). For the last ~13 hours I found 996 unique IP addresses. Only 3 of those requested more than one (i.e. 2-5) different avatars. The number of requests from each IP is very diverse though, I've found anything between 1 and 200k.

Unfortunately I don't have logs going back further since these requests cause many Gb of log files and logrotate kicks in.

Yesterday a user of QuestCraft shared their IP address with me and I saw a spike of UnityWebRequests rush in for their IP and avatar when they told me they opened QuestCraft. As such I'm pretty confident this is indeed caused by QuestCraft.

I don't think I have much more data that I can share about the traffic peak, but I know that blocking these requests were present when the server went down, and blocking them has prevented the server from becoming unavailable. Perhaps more users are using QuestCraft currently, or something else caused the requests to become this aggressive?

**Comment by CADIndie** [comment]
I'm honestly not sure, we'll take a deeper look into this and do some more in depth debugging but the only thing I can suspect as of right now is that its a bug in Unity some how. By the way I have added the "QuestCraft" user agent and it'll be used included in the next release which will hopefully be next week however we will hold off until we can ensure this issue is fixed.

**Comment by thejudge156** [comment]
I believe I have fixed this, doing more testing now.

(QuestCraft head dev btw)

**Comment by CADIndie** [comment]
The issue should now be fixed, we will release the update soon. Please DM us if anymore issues persist, and we once again do apologize for the inconvience.

**Comment by jomo** [comment]
Thanks for addressing this. FYI I have disabled the redirect and now return a 429 in order not to cause unnecessary requests to github.com.

@thejudge156 would you mind sharing some technical details about the bug and how you fixed it? I'm interested!

**Comment by CADIndie** [comment]
@jomo We believe that when the user was signing, it would loop the code that would make requests to Craftar. Why it only just became an issue, we honestly have no clue why but now its only requested once and we have also implemented the "QuestCraft" User-Agent so if there is still issues caused by us please feel free to contact us at any time and pause our User-Agent if required.

> @thejudge156 would you mind sharing some technical details about the bug and how you fixed it? I'm interested!

**Comment by jomo** [comment]
This problem does not seem to be fixed. While some clients are now using the `QuestCraft` User-Agent, they still send dozens ­— sometimes several hundred — requests per second of the same URL.

**Comment by thejudge156** [comment]
Is this still occuring?

**Comment by jomo** [comment]
Yes, the same thing is still happening.

**Comment by thejudge156** [comment]
Alright, we will make sure to look over this some more

**Comment by Tobias-Auer** [comment]
Is there anything new in the meantime 🫣, because the servers are down again :(

**Comment by jomo** [comment]
As this is still not fixed half a year later even with the `QuestCraft` User-Agent now supplied, I am redirecting those requests here. Please fix your application to not be a DDoS botnet and provide the version number in the User-Agent string if you want to keep using crafatar.com.

**Comment by CADIndie** [comment]
Oh? We checked with a local version and the same issue wasn't happening on our end, hence us not pushing the issue further. I would also like to ask, does craftar.com not have rate limiting?

**Comment by jomo** [comment]
There is no rate limit for legitimate requests, I rely on users to not spam the service. Some IP addresses identifying as QuestCraft still send several hundret requests for the same avatar per second.

**Comment by CADIndie** [comment]
We believe we have identified why and will test in a moment, as for the rate limit, even if issues such as this are rare I still really recommend you implement at least base level rate limiting to avoid issues like this in the future. Of course, that is not to say you are to blame, we apologize for the spam and have made this top priority.

**Comment by CADIndie** [comment]
This issue has been resolved and no more requests should come Crafatar's way, however it is recommended you still block the previous UA's. We once again apologize for the previous issues and would also like to emphasise once more that basic rate limiting is almost a requirement, with this being a public issue it'd be really easy for people to abuse it on purpose and in a manor magnitudes worse so its highly recommended you take care of that ASAP.
